### PR TITLE
esp32: Support building with ESP IDF 4.0-beta1

### PR DESCRIPTION
**Pull request by jimmo**
Adds support for a second supported hash (currently set to the 4.0-beta1 tag). When this hash is detected, the relevant changes are applied.

This allows us to start using v4 features (e.g. BLE with Nimble), and also start doing testing.

Note: this feature is experimental, not well tested, and network.LAN and network.PPP are currently unsupported.

Ref #5070

### CODE REVIEW

**Comment by dpgeorge** [comment]
Thanks, looks really good, a neat way to support the "old" and "new" IDF, something which I think is important: to be able to retain stability for those who need it, and also start using new features.

Can you please update the .travis.yml file to build esp32 for both v3 and v4?  Note the use of `$(grep "ESPIDF_SUPHASH :=" ports/esp32/Makefile | cut -d " " -f 3)` which will need to be updated.

**Comment by jimmo** [comment]
> Can you please update the .travis.yml file to build esp32 for both v3 and v4?

Done.

**Comment by MrSurly** [comment]
> and network.LAN and network.PPP are currently unsupported.

Hi @jimmo,

I authored these for ESP32 -- did you need me to take a stab at making them work?  Was it a matter of time, or is the API very different?

**Comment by MrSurly** [comment]
I'm seeing conditional compilation based on whether we're using V4 in the C code,  but changes to stuff like the `sdkconfig.*` don't appear to be conditional.  Are these changes compatible with both V3 and V4?

**Comment by jimmo** [comment]
> I authored these for ESP32 -- did you need me to take a stab at making them work? Was it a matter of time, or is the API very different?

It was just to get this building, I haven't even looked into what the difference is, I imagine it's quite minor. I was going to have a quick look today, but what would actually be super useful is if you can test them as I do not have the requisite hardware. Thank you!

> I'm seeing conditional compilation based on whether we're using V4 in the C code, but changes to stuff like the sdkconfig.* and don't appear to be conditional. Also a couple of places in Makefile. Are these changes compatible with both V3 and V4?

Yes. In the case of sdkconfig.* any config variables that were renamed I just provide both (with the 3.x ones at the bottom). I did not find any cases where the variables' meaning changed. In the Makefile, the cases are additional include/lib dirs (especially ones found using `wildcard`) where they will just be ignored on 3.x builds.

File: .travis.yml [hunk 1]
```diff
@@ -145,7 +145,24 @@ jobs:
...
+    - stage: test
+      env: NAME="esp32 port build with IDF v4"
+      install:
+        - sudo apt-get install python3-pip
+        - sudo pip3 install pyparsing
+        - wget https://dl.espressif.com/dl/xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz
+        - zcat xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz | tar x
+        - export PATH=$(pwd)/xtensa-esp32-elf/bin:$PATH
+        - git clone https://github.com/espressif/esp-idf.git
+        - git -C esp-idf checkout $(grep "ESPIDF_SUPHASH_V4 :=" ports/esp32/Makefile | cut -d " " -f 3)
```
Comment (dpgeorge) [inline .travis.yml]: I'd suggest to have just one "stage" for both V3 and V4 IDF builds, to make the overall build faster.  Something like this:
```
    # esp32 port
    - stage: test
      env: NAME="esp32 port build"
      install:
        - sudo apt-get install python3-pip
        - sudo pip3 install pyparsing
        - wget https://dl.espressif.com/dl/xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz
        - zcat xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz | tar x
        - export PATH=$(pwd)/xtensa-esp32-elf/bin:$PATH
        - git clone https://github.com/espressif/esp-idf.git
        - export IDF_PATH=$(pwd)/esp-idf
      script:
        - git submodule update --init lib/berkeley-db-1.xx
        - make ${MAKEOPTS} -C mpy-cross
        # IDF v3 build
        - git -C esp-idf checkout $(grep "ESPIDF_SUPHASH_V3 :=" ports/esp32/Makefile | cut -d " " -f 3)
        - git -C esp-idf submodule update --init components/json/cJSON components/esp32/lib components/esptool_py/esptool components/expat/expat components/lwip/lwip components/mbedtls/mbedtls components/micro-ecc/micro-ecc components/nghttp/nghttp2
        - make ${MAKEOPTS} -C ports/esp32
        # clean
        - git -C esp-idf clean -f -f -d components/json/cJSON components/esp32/lib components/expat/expat components/micro-ecc/micro-ecc components/nghttp/nghttp2
        - make ${MAKEOPTS} -C ports/esp32 clean
        # IDF v4 build
        - git -C esp-idf checkout $(grep "ESPIDF_SUPHASH_V4 :=" ports/esp32/Makefile | cut -d " " -f 3)
        - git -C esp-idf submodule update --init components/esp_wifi/lib_esp32 components/esptool_py/esptool components/lwip/lwip components/mbedtls/mbedtls
        - make ${MAKEOPTS} -C ports/esp32
```

File: ports/esp32/Makefile [hunk 2]
```diff
@@ -136,13 +147,28 @@ INC_ESPCOMP += -I$(ESPCOMP)/vfs/include
...
+INC_ESPCOMP += -I$(ESPCOMP)/spi_flash/private_include
+INC_ESPCOMP += -I$(ESPCOMP)/xtensa/include
+INC_ESPCOMP += -I$(ESPCOMP)/xtensa/esp32/include
+INC_ESPCOMP += -I$(ESPCOMP)/esp_rom/include
+INC_ESPCOMP += -I$(ESPCOMP)/esp_common/include
+INC_ESPCOMP += -I$(ESPCOMP)/esp_wifi/include
+INC_ESPCOMP += -I$(ESPCOMP)/esp_wifi/esp32/include
+else
+INC_ESPCOMP += -I$(ESPCOMP)/ethernet/include
+endif
```
Comment (dpgeorge) [inline ports/esp32/Makefile]: The expat, micro-ecc and nghttp lines (5 in total from above) will also need to be moved here.

File: ports/esp32/Makefile [hunk 3]
```diff
@@ -382,18 +423,44 @@ ESPIDF_LWIP_O = $(patsubst %.c,%.o,\
...
+ifeq ($(ESPIDF_CURHASH),$(ESPIDF_SUPHASH_V4))
+ESPIDF_ESP_COMMON_O = $(patsubst %.c,%.o,$(wildcard $(ESPCOMP)/esp_common/src/*.c))
+
+ESPIDF_ESP_EVENT_O = $(patsubst %.c,%.o,$(wildcard $(ESPCOMP)/esp_event/*.c))
+
+ESPIDF_ESP_WIFI_O = $(patsubst %.c,%.o,$(wildcard $(ESPCOMP)/esp_wifi/src/*.c))
+
+$(BUILD)/$(ESPCOMP)/esp_eth/src/esp_eth_mac_dm9051.o: CFLAGS += -fno-strict-aliasing
+ESPIDF_ESP_ETH_O = $(patsubst %.c,%.o,$(wildcard $(ESPCOMP)/esp_eth/src/*.c))
+
```
Comment (dpgeorge) [inline ports/esp32/Makefile]: Need to move `ESPIDF_JSON_O` here as well.

File: ports/esp32/Makefile [hunk 4]
```diff
@@ -451,6 +527,31 @@ LIB = $(foreach lib,$(LIB_ESPIDF),$(BUILD_ESPIDF_LIB)/$(lib)/lib$(lib).a)
 # ESP IDF ldgen
 
 LDGEN_FRAGMENTS = $(shell find $(ESPCOMP) -name "*.lf")
+
+ifeq ($(ESPIDF_CURHASH),$(ESPIDF_SUPHASH_V4))
+
+LDGEN_LIBRARIES=$(foreach lib,$(LIB_ESPIDF),$(BUILD_ESPIDF_LIB)/$(lib)/lib$(lib).a)
+
+$(BUILD_ESPIDF_LIB)/ldgen_libraries: $(LDGEN_LIBRARIES) $(IDF_PATH)/make/ldgen.mk
```
Comment (dpgeorge) [inline ports/esp32/Makefile]: IDF_PATH should be ESPIDF

File: ports/esp32/Makefile [hunk 5]
```diff
@@ -553,7 +661,14 @@ $(BUILD)/%.o: %.cpp
 BOOTLOADER_LIB_DIR = $(BUILD)/bootloader
 BOOTLOADER_LIB_ALL =
 
+ifeq ($(ESPIDF_CURHASH),$(ESPIDF_SUPHASH_V4))
+$(BUILD)/bootloader/$(ESPCOMP)/%.o: CFLAGS += -DBOOTLOADER_BUILD=1 -I$(ESPCOMP)/bootloader_support/include_priv -I$(ESPCOMP)/bootloader_support/include -I$(ESPCOMP)/micro-ecc/micro-ecc -I$(ESPCOMP)/efuse/include -I$(ESPCOMP)/esp_rom/include -Wno-error=format \
```
Comment (dpgeorge) [inline ports/esp32/Makefile]: Remove micro-ecc include from this line.

Comment (dpgeorge) [reply ports/esp32/Makefile]: Also, the `# libmicro-ecc.a` section below needs to be wrapped in `ifeq ($(ESPIDF_CURHASH),$(ESPIDF_SUPHASH_V3))`

Comment (dpgeorge) [reply ports/esp32/Makefile]: Also the `BOOTLOADER_LIBS += -L$(ESPCOMP)/esp32/lib -lrtc` line needs to be changed to the following for V4 (so conditional):
```
BOOTLOADER_LIBS += -L$(ESPCOMP)/esp_wifi/lib_esp32 -lrtc
```

**Comment by jimmo** [comment]
OK I've updated the branch with those comments addressed.

Will leave LAN/PPP for a follow-up PR.

File: .travis.yml [hunk 6]
```diff
@@ -145,12 +145,21 @@ jobs:
         - zcat xtensa-esp32-elf-linux64-1.22.0-80-g6c4433a-5.2.0.tar.gz | tar x
         - export PATH=$(pwd)/xtensa-esp32-elf/bin:$PATH
         - git clone https://github.com/espressif/esp-idf.git
-        - git -C esp-idf checkout $(grep "ESPIDF_SUPHASH :=" ports/esp32/Makefile | cut -d " " -f 3)
-        - git -C esp-idf submodule update --init components/json/cJSON components/esp32/lib components/esptool_py/esptool components/expat/expat components/lwip/lwip components/mbedtls/mbedtls components/micro-ecc/micro-ecc components/nghttp/nghttp2
+        - export IDF_PATH=$(pwd)/esp-idf
```
Comment (dpgeorge) [inline .travis.yml]: Looks like the `sudo pip3 install pyparsing` line above needs to be `sudo pip3 install 'pyparsing<2.4'`

Comment (jimmo) [reply .travis.yml]: Thanks, updated the branch.

**Comment by jimmo** [comment]
> I authored these for ESP32 -- did you need me to take a stab at making them work? Was it a matter of time, or is the API very different?

@MrSurly it does look like there are a few changes and I don't think I'll be able to get to it this week. If you would be able to take a look at this in a follow-up PR that would be fantastic. Thanks!

**Comment by dpgeorge** [comment]
Ok, merged in 96008ff59a8af9883af17d01b951029d9d02eec9.

File: ports/esp32/Makefile [hunk 7]
```diff
@@ -59,29 +65,51 @@ ESPIDF = $(IDF_PATH)
 else
 $(info The ESPIDF variable has not been set, please set it to the root of the esp-idf repository.)
 $(info See README.md for installation instructions.)
-$(info Supported git hash: $(ESPIDF_SUPHASH))
 $(error ESPIDF not set)
+print_supported_git_hash
```
Comment (andrewleech) [inline ports/esp32/Makefile]: This doesn't print the supported hash anymore because this line is after the error line

**Pull request closed without merging**